router/factory.go

```diff
@@ -1,26 +1,30 @@
 package router
 
 import (
+	"sync"
+
+	"github.com/go-redis/redis/v8"
+
+	"github.com/rudderlabs/rudder-server/config"
 	backendconfig "github.com/rudderlabs/rudder-server/config/backend-config"
+	"github.com/rudderlabs/rudder-server/internal/throttling"
 	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/router/throttler"
 	"github.com/rudderlabs/rudder-server/services/rsources"
+	"github.com/rudderlabs/rudder-server/services/stats"
 	"github.com/rudderlabs/rudder-server/services/transientsource"
+	"github.com/rudderlabs/rudder-server/utils/logger"
 )
 
-type destinationConfig struct {
-	name          string
-	responseRules map[string]interface{}
-	config        map[string]interface{}
-	destinationID string
-}
+const (
+	throttlingAlgoTypeGoRate         = "gorate"
+	throttlingAlgoTypeGCRA           = "gcra"
+	throttlingAlgoTypeRedisGCRA      = "redis-gcra"
+	throttlingAlgoTypeRedisSortedSet = "redis-sorted-set"
+)
 
-func getRouterConfig(destination *backendconfig.DestinationT, identifier string) destinationConfig {
-	return destinationConfig{
-		name:          destination.DestinationDefinition.Name,
-		destinationID: identifier,
-		config:        destination.DestinationDefinition.Config,
-		responseRules: destination.DestinationDefinition.ResponseRules,
-	}
+type limiter interface {
+	CheckLimitReached(key string, cost int64) (limited bool, retErr error)
 }
 
 type Factory struct {
```

Comment:
By providing a producer function for `Factory` and always creating factories using this function, we could perform initialisation once, without using a `sync.Once` and having to call it first thing at the beginning of each struct's method.